File: Algorithms/075_SortColors.java
public class Solution {
    public void sortColors(int[] A) {
        int i=0, j=A.length-1, temp=0;
        while (i < A.length && A[i] == 0) {
        	i++;
        }
        while (j >= 0 && A[j] == 2) {
        	j--;
        }
        for (int k = i; k <= j; ++k)
        {
        	if (A[k] == 0) {
        		temp = A[i];
        		A[i] = A[k];
        		A[k] = temp;
        		while (A[i] == 0) {
        			i++;
        		}
	        	if (A[k] == 2) {
	        		temp = A[j];
	        		A[j] = A[k];
	        		A[k] = temp;
	        		while (A[j] == 2) {
	        			j--;
	        		}
	        	}
        	}
        	if (A[k] == 2) {
        		temp = A[j];
        		A[j] = A[k];
        		A[k] = temp;
        		while (A[j] == 2) {
        			j--;
        		}
        		if (A[k] == 0) {
        			temp = A[i];
        			A[i] = A[k];
        			A[k] = temp;
        			while (A[i] == 0) {
        				i++;
        			}
        		}
        	}
        	if (k < i) {
        		k = i;
        	}
        }        
    }
}

File: Algorithms/py/075.SortColors.py
class Solution(object):
    def sortColors(self, nums):
        """
        :type nums: List[int]
        :rtype: void Do not return anything, modify nums in-place instead.
        """
        for i in range(len(nums)):
            if nums[i] != 0:
                break
        for j in range(len(nums)-1, -1, -1):
            if nums[j] != 2:
                break
        k = i
        while k <= j:
            if nums[k] == 0:
                nums[i], nums[k] = nums[k], nums[i]
                i += 1
            elif nums[k] == 2:
                nums[j], nums[k] = nums[k], nums[j]
                j -= 1
            else: k += 1
            if k <= i:
                k = i
